opentelemetry-sdk/src/logs/log_processor.rs

```diff
@@ -994,4 +994,232 @@ mod tests {
+            for (log_record, instrumentation) in batch.iter() {
+                logs_lock.push((log_record.clone(), instrumentation.clone()));
+            }
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn test_simple_processor_async_exporter_without_runtime() {
+        // Use `catch_unwind` to catch the panic caused by missing Tokio runtime
+        let result = std::panic::catch_unwind(|| {
+            let exporter = LogExporterThatRequiresTokioSpawn::new();
+            let processor = SimpleLogProcessor::new(Box::new(exporter.clone()));
+
+            let mut record: LogRecord = Default::default();
+            let instrumentation: InstrumentationLibrary = Default::default();
+
+            // This will panic because `tokio::spawn` or an async operation is called without a runtime.
+            processor.emit(&mut record, &instrumentation);
+        });
+
+        // Verify that the panic occurred and check the panic message for the absence of a Tokio runtime
+        assert!(
+            result.is_err(),
+            "The test should fail due to missing Tokio runtime, but it did not."
+        );
+
+        if let Err(panic) = result {
+            let panic_message = panic
+                .downcast_ref::<String>()
+                .map(|s| s.as_str())
+                .or_else(|| panic.downcast_ref::<&str>().copied())
+                .unwrap_or("No panic message");
+
+            assert!(
+                panic_message.contains("no reactor running")
+                    || panic_message
+                        .contains("must be called from the context of a Tokio 1.x runtime"),
+                "Expected panic message about missing Tokio runtime, but got: {}",
+                panic_message
+            );
+        }
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+    async fn test_simple_processor_async_exporter_with_runtime() {
+        let exporter = LogExporterThatRequiresTokioSpawn::new();
+        let processor = SimpleLogProcessor::new(Box::new(exporter.clone()));
+
+        let mut record: LogRecord = Default::default();
+        let instrumentation: InstrumentationLibrary = Default::default();
+
+        processor.emit(&mut record, &instrumentation);
+
+        assert_eq!(exporter.len().await, 1);
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+    #[ignore]
+    // all threads are blocked running blocked::export(), and the exporter further needs tokio
+    // runtime to progress on this blocked thread, resulting in deadlock.
+    async fn test_simple_processor_async_exporter_with_multi_thread_runtime_all_cores_blocked() {
+        let exporter = LogExporterThatRequiresTokioSpawn::new();
+        let processor = Arc::new(Mutex::new(SimpleLogProcessor::new(Box::new(
+            exporter.clone(),
+        ))));
+
+        let concurrent_emit = 5; // number of worker sthreads + 1
+
+        let mut handles = vec![];
+        // send 2 events concurrently
```

Comment:
not sure I follow this.. each task take a mutex lock to invoke emit on processor, so processors are called serially only right?

Reply:
Yes the processor::emit is called serially. However this does spawn's  4 out of 5 requested async tasks (on 4 available worker threads) concurrently, one task is able to move forward, and invoke exporter::export() method. The rest of the tokio tasks are blocked waiting for the processor lock. So the running exporter task doesn't get any free worker thread to run the tokio's sleep method, resulting in deadlock.
This mimics the scenario where use invokes tracing event concurrently from max available worker threads, and all but one wait on the SimpleLogProcessor export lock to be released, resulting in the deadlock.

Reply:
This is somehow to demonstrate that even in case the tokio::main is invoked in multiple thread mode, there is possibility of deadlock just because the futures::block_on ensures any async calls inside it run on the current worker thread.